### PR TITLE
Added support for dynamic text/string in RegexpString() and String() …

### PR DESCRIPTION
**Pull request by jshenguru**
In conjunction with [this replace-response PR](https://github.com/caddyserver/replace-response/pull/33/commits/1596637dc22b1d450bb3d77a66cead8645097f96). Details are discussed [here](https://github.com/caddyserver/replace-response/issues/26#issuecomment-2518534936).

### CODE REVIEW

File: replace.go [hunk 1]
```diff
@@ -35,8 +35,8 @@ func Bytes(old, new []byte) Transformer {
 
 // String returns a transformer that replaces all instances of old with new.
 // Unlike strings.Replace, empty old values don't match anything.
-func String(old, new string) Transformer {
-	return Bytes([]byte(old), []byte(new))
+func String(old, new string, replace func(string) string) Transformer {
```
Comment (icholy) [inline replace.go:38]: This should be a new function call `StringFunc`.

File: replace.go [hunk 2]
```diff
@@ -126,9 +126,9 @@ func Regexp(re *regexp.Regexp, new []byte) *RegexpTransformer {
 
 // RegexpString returns a transformer that replaces all matches of re with template
 // Inside template, $ signs are interpreted as in Expand, so for instance $1 represents the text of the first submatch.
-func RegexpString(re *regexp.Regexp, template string) *RegexpTransformer {
+func RegexpString(re *regexp.Regexp, template string, replace func(string) string) *RegexpTransformer {
 	return RegexpIndexFunc(re, func(src []byte, index []int) []byte {
-		return re.Expand(nil, []byte(template), src, index)
+		return re.Expand(nil, []byte(replace(template)), src, index)
```
Comment (icholy) [inline replace.go:131]: This change makes no sense. It should be:

``` go
return re.Expand(nil, []byte(replace(string(src))), src, index)
```

And it should be a new function called `RegexpStringFunc`.

edit:

nvm this change doesn't make sense either.

File: replace.go [hunk 3]
```diff
@@ -35,8 +35,8 @@ func Bytes(old, new []byte) Transformer {
 
 // String returns a transformer that replaces all instances of old with new.
 // Unlike strings.Replace, empty old values don't match anything.
-func String(old, new string) Transformer {
-	return Bytes([]byte(old), []byte(new))
+func String(old, new string, replace func(string) string) Transformer {
+	return Bytes([]byte(replace(old)), []byte(replace(new)))
```
Comment (icholy) [inline replace.go:39]: What's the purpose of this change?

**Comment by icholy** [comment]
Thanks for the PR, but I'll have to decline this change for several reasons:

1. This is a breaking change to the public API that would require all existing users to modify their code, yet it doesn't provide any new functionality that wasn't already possible.

2. If a user needs to transform the strings before creating a transformer, they can simply do that at the call site:

   ```go
   // Instead of:
   String(old, new, myReplace)

   // Users can already do:
   String(myReplace(old), myReplace(new))
   ```

3. I'm confused about the intent of the `RegexpString` change, why is the `template` string being passed to a replace function?

If you have a specific use case that isn't well-served by the current API, I'd be happy to discuss it further to find a solution that doesn't break compatibility.

**Comment by mholt** [comment]
Hmm, I thought the intent was to allow the strings to be generated _after_ the creation of the transformer but before the transformations occur. Our goal is to avoid re-allocating transformers for every HTTP request (in our case).

Maybe there's a non-breaking API that could be introduced, like new methods/functions?

**Comment by jshenguru** [comment]
@icholy and @mholt I just created new PRs to both icholy/replace (https://github.com/icholy/replace/pull/8) and caddyserver/replace-response (https://github.com/caddyserver/replace-response/pull/34). I introduced two new Transformer methods to the icholy/replace to keep the already-published APIs intact. Also explained the reasons in the comment lines of the new methods. Also named the new methods with more semantics. Please review

**Comment by icholy** [comment]
> the intent was to allow the strings to be generated after the creation of the transformer but before the transformations occur. 

@mholt this PR doesn't implement anything like that though.

**Comment by mholt** [comment]
@icholy I see. Thanks for clarifying. I have not followed it very close.
